programs/marginfi/src/state/native_oracle/pyth_crosschain.rs

```diff
@@ -0,0 +1,231 @@
+    pub ema_price: i64,
+    pub ema_conf: u64,
+    pub min_verificaiton_level: VerificationLevel,
+}
+
+impl PythnetPriceFeed {
+    /// Create a new PythnetPriceFeed from a PriceUpdateV2
+    /// Checks:
+    /// - Verification level is sufficient
+    ///
+    /// Does not check for staleness, or feed ID matching
+    ///
+    /// Staleness is ignored because this is the first time we are setting up the feed,
+    /// getting a live feed is not practical.
+    ///
+    /// Feed ID check is done in the `validate_oracle_setup` call.
+    pub fn new_no_staleness_check(
+        price_update: PriceUpdateV2,
+        min_verificaiton_level: VerificationLevel,
+    ) -> MarginfiResult<Self> {
+        let PriceFeedMessage {
+            price,
+            conf,
+            exponent,
+            publish_time,
+            ema_price,
+            ema_conf,
+            ..
+        } = price_update.price_message;
+
+        check!(
+            price_update.verification_level.gte(min_verificaiton_level),
+            MarginfiError::IllegalOracleUpdate,
+            "Insufficient verification level"
+        );
+
+        Ok(Self {
+            price,
+            conf,
+            exponent,
+            publish_time,
+            ema_price,
+            ema_conf,
+            min_verificaiton_level,
+        })
+    }
+
+    /// Update the PythnetPriceFeed with a new PriceUpdateV2
+    ///
+    /// Checks:
+    /// - Feed ID matches the expected feed ID
+    /// - Verification level is sufficient
+    /// - Oracle is not stale
+    /// - Update is newer than the current price update
+    pub fn try_update(
+        &mut self,
+        new_price_update: &PriceUpdateV2,
+        ctl: PythnetPriceFeedControl,
+    ) -> MarginfiResult<()> {
+        let PriceFeedMessage {
+            ref feed_id,
+            price,
+            conf,
+            exponent,
+            publish_time,
+            ema_price,
+            ema_conf,
+            ..
+        } = new_price_update.price_message;
+
+        ctl.check_feed_id_matches(feed_id)?;
+        ctl.check_sufficient_verificaiton_level(new_price_update.verification_level)?;
+
+        check!(
+            publish_time > self.publish_time,
+            MarginfiError::IllegalOracleUpdate,
+            "Update is older than current price update"
+        );
+
+        *self = Self {
+            price,
+            conf,
+            exponent,
+            publish_time,
+            ema_price,
+            ema_conf,
+            min_verificaiton_level: self.min_verificaiton_level,
+        };
+
+        Ok(())
+    }
+
+    pub fn check_staleness(&self, current_time: i64, max_age: u64) -> MarginfiResult<()> {
+        check!(
+            self.publish_time.saturating_add(max_age as i64) >= current_time,
+            MarginfiError::StaleOracle,
+            "Oracle is stale"
+        );
+
+        Ok(())
+    }
+
+    #[inline(always)]
+    fn get_ema_price(&self) -> MarginfiResult<I80F48> {
```

Comment:
may want to add unweighted prefix to be consistent with `get_unweighted_price`:


```suggestion
    fn get_unweighted_ema_price(&self) -> MarginfiResult<I80F48> {
```